### PR TITLE
[StableHLO] Support 3D scaled_dot_product_attention in lowering and sharding rule

### LINKED ISSUE #5914: [HunyuanImage-2.1-Distilled-Diffusers] Decoder faces 'ttir.scaled_dot_product_attention' op Query must be a 4D tensor error
### Problem description 

- `tests/torch/models/HunyuanImage_2_1/test_vae_decoder.py::test_vae_decoder` in bfloat16 fails with `error: 'ttir.scaled_dot_product_attention' op Query must be a 4D tensor` 
- Issue originates from this [SDPA](https://github.com/huggingface/diffusers/blob/a3608b512ed7248499a44c61d954965ed9bdae4d/src/diffusers/models/autoencoders/autoencoder_kl_hunyuanimage.py#L113) whose inputs are 3D 

### Logs

- [aug7_HunyuanImage_2_1_decoder.log](https://github.com/user-attachments/files/30814793/aug7_HunyuanImage_2_1_decoder.log), 
- [test_vae_decoder.py](https://github.com/user-attachments/files/30814794/test_vae_decoder.py)
- [aug10_sanity_mc_bf.log](https://github.com/user-attachments/files/30895054/aug10_sanity_mc_bf.log)
- [aug10_sanity_sc_bf.log](https://github.com/user-attachments/files/30895053/aug10_sanity_sc_bf.log)
- [test_sanity_mc.py](https://github.com/user-attachments/files/30895051/test_sanity_mc.py)
- [test_sanity.py](https://github.com/user-attachments/files/30895052/test_sanity.py)

### PR DESCRIPTION
**Pull request by kamalrajkannan78**
## Ticket
- fixes https://github.com/tenstorrent/tt-xla/issues/5914

## Problem Description

- The `stablehlo.composite "tenstorrent.scaled_dot_product_attention"` lowering passes query/key/value through unchanged. `torch.nn.functional.scaled_dot_product_attention` accepts rank-3 `(B, S, E)` input, but `ttir.scaled_dot_product_attention` requires 4D `[B, Hq, Sq, D]`, so single-head attention blocks that never materialize a head dimension fail to convert.
- HunyuanImage 2.1's VAE decoder mid-block attention is single-head and reshapes to `(B, H*W, C)` before calling SDPA, emitting a `1x4096x1024` composite that is rejected during SHLO -> TTIR conversion.
- On the multi-chip path the composite becomes a `stablehlo.custom_call` so Shardy can propagate through it, and `getSDPAShardingRule` only handles rank 4. Rank 3 falls back to `buildPointwise`, which lets the sequence dim be sharded.

## What's Changed

- Promote rank-3 query/key/value to 4D in `convertToTTIRScaledDotProductAttention` (covers both the composite and sharded `custom_call` paths), then reshape the result back
- The unit dim is inserted at index 1 (after batch), since `ttir.scaled_dot_product_attention` reads its shape positionally as `[B, Hq, Sq, D]`.
- A rank-3 mask is promoted the same way rather than left-padded, keeping the batch dim at index 0 (previously correct only for `B == 1`); other ranks rejected via `notifyMatchFailure`, 4D unchanged
- Added a rank-3 branch to `getSDPAShardingRule`: batch is `kPassThrough`, sequence and hidden are `kNeedReplication`, so the sequence dim is gathered instead of sharded
- Added `test/ttmlir/Conversion/StableHLOToTTIR/composite/test_sdpa_rank3.mlir`: rank-3 promotion, mask alignment at `B = 2`, 4D pass-through regression
- Added `test/ttmlir/Dialect/StableHLO/register_custom_sharding_rule/custom_op_sdpa_rank3.mlir`: batch-sharded and sequence-sharded rank-3 cases

## Checklist
- [x] Verified changes through local testing on sanity & model in BH

### Logs

- [aug7_HunyuanImage_2_1_decoder_before_fix.log](https://github.com/user-attachments/files/30822212/aug7_HunyuanImage_2_1_decoder.log)
- [aug10_sanity_mc_before_fix.log](https://github.com/user-attachments/files/30895054/aug10_sanity_mc_bf.log)
- [aug10_sanity_sc_before_fix.log](https://github.com/user-attachments/files/30895053/aug10_sanity_sc_bf.log)
- [aug10_HunyuanImage_2_1_decoder_after_fix.log](https://github.com/user-attachments/files/30895254/aug10_HunyuanImage_2_1_decoder_after_fix.log)
- [aug10_sanity_mc_after_fix.log](https://github.com/user-attachments/files/30895256/aug10_sanity_mc_af.log)
- [aug10_sanity_sc_after_fix.log](https://github.com/user-attachments/files/30895257/aug10_sanity_sc_af.log)

### CODE REVIEW

File: lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp [hunk 1]
```diff
@@ -1707,20 +1708,54 @@ static LogicalResult convertToTTIRScaledDotProductAttention(
     }
   }
 
+  auto queryType = mlir::cast<RankedTensorType>(operands[0].getType());
+  int64_t queryRank = queryType.getRank();
+  if (queryRank != 3 && queryRank != 4) {
+    return rewriter.notifyMatchFailure(srcOp, "query rank must be 3 or 4");
```
Comment (sshonTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp:1714]: https://github.com/tenstorrent/tt-mlir/blob/main/lib/Dialect/StableHLO/Transforms/RegisterCustomShardingRule.cpp#L189

Could you also check the sharding rule part? Now that this PR supports 3D, we need a rule for 3D as well. We definitely don't want PointWise here. Also, please add a test for this new rule.

Comment (kamalrajkannan78) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp:1714]: Good catch, done. Added a rank-3 branch to `getSDPAShardingRule`: batch is `kPassThrough`, sequence and hidden are `kNeedReplication`.

Confirmed on 4 chips — with the pointwise fallback Shardy sharded the sequence dim (`1x1x4096x1024` -> `1x1x1024x1024`) and the multichip sanity test failed with `pcc=0.498`. With the rule the sequence is gathered back to full length and it passes. 

- [aug10_sanity_mc_pointwise_fallback.log](https://github.com/user-attachments/files/30895369/aug10_sanity_mc_af.log)
- [aug10_sanity_mc_after_fix.log](https://github.com/user-attachments/files/30895370/aug10_sanity_mc_asf.log)


Test added in `test/ttmlir/Dialect/StableHLO/register_custom_sharding_rule/custom_op_sdpa_rank3.mlir`, covering batch-sharded and sequence-sharded rank-3 cases.

Comment (kamalrajkannan78) [reply lib/Conversion/StableHLOToTTIR/StableHLOLegalizeCompositePass.cpp:1714]: updated the PR description and logs to reflect current changes

**Review by sshonTT — APPROVED** [review]
